src/Chainweb/Pact/Backend/PactState/GrandHash.hs

```diff
@@ -0,0 +1,753 @@
+  --   2) In the third phase, we pair all of the grand hashes with the associated
+  --      block headers.
+
+  -- Phase 1:
+  --   Check that the DB matches our expectations and resolve
+  --   any slight mismatches between chains
+  chainTargets <- resolveTargets logger cids pactDir ts
+
+  -- Phase 2:
+  --   Gather all of the grand hashes
+
+  -- This holds the Grand Hashes for each chain, at each BlockHeight of
+  -- interest.
+  chainHashesRef <- newIORef @(Map BlockHeight (Map ChainId ByteString)) Map.empty
+
+  -- It's simpler to instantiate the keys ahead of time
+  --forM_ targets $ \target -> do
+  --  modifyIORef' chainHashesRef $ \m -> Map.insert target Map.empty m
+
+  pooledFor cids $ \cid -> do
+    withChainDb cid logger pactDir $ \(SQLiteEnv db _) -> do
+      case Map.lookup cid chainTargets of
+        Nothing -> do
+          pure ()
+        Just targets -> do
+          -- Compute the grand hash for the chain at every target height
+          pooledFor targets $ \target -> do
+            hash <- computeGrandHash db target
+            atomicModifyIORef' chainHashesRef $ \m ->
+              (Map.insertWith Map.union target (Map.singleton cid hash) m, ())
+
+  -- Note the toDescList. This is so that pact-import's top-down search
+  -- is faster, and correct.
+  chainHashes <- Map.toDescList <$> readIORef chainHashesRef
+
+  -- Phase 3:
+  -- Grab the headers corresponding to every blockheight from RocksDB
+  withReadOnlyRocksDb rocksDir modernDefaultOptions $ \rocksDb -> do
+    forM chainHashes $ \(height, hashes) -> do
+      headers <- getBlockHeadersAt rocksDb chainTargets height chainwebVersion
+      let missingIn cid m = error $ "missing entry for chain " <> Text.unpack (chainIdToText cid) <> " in " <> m
+      pure $ (height,)
+        $ Merge.merge
+            (Merge.mapMissing (\cid _ -> missingIn cid "headers"))
+            (Merge.mapMissing (\cid _ -> missingIn cid "hashes"))
+            (Merge.zipWithMatched (\_ hash header -> (hash, header)))
+            hashes
+            headers
+
+data PactCalcConfig = PactCalcConfig
+  { pactDir :: FilePath
+  , rocksDir :: FilePath
+  , chainwebVersion :: ChainwebVersion
+  , targetBlockHeight :: BlockHeightTargets
+  }
+
+-- | Calculate the hash at every provided blockheight across all chains.
+--
+--   Note that for some chains, one or more of the requested blockheights
+--   won't be accessible. This could be due to compaction, or the blockheight
+--   predating the genesis of the chain. In this case, you will receive a
+--   warning.
+pactCalcMain :: IO ()
+pactCalcMain = do
+  cfg <- O.execParser opts
+  let cids = allChains cfg.chainwebVersion
+  checkPactDbsExist cfg.pactDir cids
+
+  C.withDefaultLogger YAL.Info $ \logger -> do
+    chainHashes <- computeGrandHashesAt logger cids cfg.pactDir cfg.rocksDir cfg.targetBlockHeight cfg.chainwebVersion
+    writeFile "src/Chainweb/Pact/Backend/PactState/EmbeddedHashes.hs" (chainHashesToModule chainHashes)
```

Comment:
This tool should be run from the root of the chainweb-node repo. Not sure if there's a good way to enforce this.
We should probably catch an exception here that says the path does not exist.